File: ios/Algo/Features/Notifications/ViewModels/NotificationsViewModel.swift
// NotificationsViewModel.swift
// Algo
//
// ViewModel for the Notifications screen. Manages notification list,
// filtering, read status, and pagination.

import Foundation

// MARK: - NotificationFilter

/// Filter options for the notification list.
enum NotificationFilter: String, CaseIterable, Identifiable {
    case all = "All"
    case unread = "Unread"

    var id: String { rawValue }
}

// MARK: - NotificationsViewModel

/// Manages the state and business logic for the Notifications screen.
///
/// Responsibilities:
/// - Fetches paginated notifications from the backend
/// - Filters by all vs. unread
/// - Marks individual notifications as read
/// - Marks all notifications as read
/// - Tracks unread count for the tab badge
@MainActor
final class NotificationsViewModel: ObservableObject {

    // MARK: Published State

    /// The current list of notifications.
    @Published private(set) var notifications: [AppNotification] = []

    /// Total count of unread notifications.
    @Published private(set) var unreadCount: Int = 0

    /// Whether a data load is in progress.
    @Published private(set) var isLoading = false

    /// Current error message, if any.
    @Published private(set) var error: String?

    /// The active filter selection.
    @Published var selectedFilter: NotificationFilter = .all {
        didSet {
            if oldValue != selectedFilter {
                Task { await loadNotifications() }
            }
        }
    }

    // MARK: Pagination

    /// Current page for pagination.
    private var currentPage = 1

    /// Whether more pages are available.
    private(set) var hasMorePages = true

    /// Whether a page load is in progress (for infinite scroll).
    @Published private(set) var isLoadingMore = false

    // MARK: Dependencies

    private let service: NotificationServiceProtocol

    // MARK: Initialization

    /// Creates a new `NotificationsViewModel`.
    /// - Parameter service: The notification service to use for API calls.
    init(service: NotificationServiceProtocol) {
        self.service = service
    }

    // MARK: Public Methods

    /// Loads the first page of notifications.
    /// Call this from `.task` on the view's appearance.
    func loadNotifications() async {
        isLoading = true
        error = nil
        currentPage = 1
        hasMorePages = true

        do {
            let response = try await service.getNotifications(
                page: 1,
                unreadOnly: selectedFilter == .unread
            )
            notifications = response.notifications
            unreadCount = response.unreadCount
            hasMorePages = response.notifications.count >= 20
        } catch {
            self.error = error.localizedDescription
        }

        isLoading = false
    }

    /// Loads the next page of notifications for infinite scroll.
    func loadMoreIfNeeded(currentItem: AppNotification) async {
        guard hasMorePages,
              !isLoadingMore,
              let lastItem = notifications.last,
              currentItem.id == lastItem.id
        else { return }

        isLoadingMore = true
        currentPage += 1

        do {
            let response = try await service.getNotifications(
                page: currentPage,
                unreadOnly: selectedFilter == .unread
            )
            notifications.append(contentsOf: response.notifications)
            unreadCount = response.unreadCount
            hasMorePages = response.notifications.count >= 20
        } catch {
            currentPage -= 1
        }

        isLoadingMore = false
    }

    /// Refreshes the notification list (pull-to-refresh).
    func refresh() async {
        await loadNotifications()
    }

    /// Marks a single notification as read and updates the local state.
    /// - Parameter notification: The notification to mark as read.
    func markAsRead(_ notification: AppNotification) async {
        guard !notification.read else { return }

        do {
            try await service.markAsRead(id: notification.id)

            // Update local state optimistically
            if let index = notifications.firstIndex(where: { $0.id == notification.id }) {
                // Create an updated copy. Since AppNotification properties are `let`,
                // we replace the item entirely.
                let updated = AppNotification(
                    id: notification.id,
                    type: notification.type,
                    title: notification.title,
                    body: notification.body,
                    severity: notification.severity,
                    read: true,
                    actionUrl: notification.actionUrl,
                    createdAt: notification.createdAt
                )
                notifications[index] = updated
                unreadCount = max(0, unreadCount - 1)
            }
        } catch {
            // Silently fail; the notification will remain as unread.
        }
    }

    /// Marks all notifications as read.
    func markAllAsRead() async {
        do {
            try await service.markAllAsRead()

            // Update all local notifications to read
            notifications = notifications.map { notification in
                AppNotification(
                    id: notification.id,
                    type: notification.type,
                    title: notification.title,
                    body: notification.body,
                    severity: notification.severity,
                    read: true,
                    actionUrl: notification.actionUrl,
                    createdAt: notification.createdAt
                )
            }
            unreadCount = 0
        } catch {
            self.error = "Failed to mark all as read."
        }
    }
}

File: android/app/src/main/java/com/algonit/algo/features/notifications/presentation/NotificationsViewModel.kt
package com.algonit.algo.features.notifications.presentation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.algonit.algo.features.notifications.data.model.AppNotification
import com.algonit.algo.features.notifications.data.model.NotificationListResponse
import dagger.hilt.android.lifecycle.HiltViewModel
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.get
import io.ktor.client.request.parameter
import io.ktor.client.request.post
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

enum class NotificationFilter {
    All,
    Unread
}

data class NotificationsUiState(
    val notifications: List<AppNotification> = emptyList(),
    val unreadCount: Int = 0,
    val isLoading: Boolean = false,
    val isRefreshing: Boolean = false,
    val error: String? = null,
    val filter: NotificationFilter = NotificationFilter.All,
    val currentPage: Int = 1,
    val hasMore: Boolean = true,
    val isLoadingMore: Boolean = false
)

@HiltViewModel
class NotificationsViewModel @Inject constructor(
    private val httpClient: HttpClient
) : ViewModel() {

    private val _uiState = MutableStateFlow(NotificationsUiState())
    val uiState: StateFlow<NotificationsUiState> = _uiState.asStateFlow()

    val unreadCount: StateFlow<Int> = MutableStateFlow(0).also { flow ->
        viewModelScope.launch {
            _uiState.collect { state ->
                (flow as MutableStateFlow).value = state.unreadCount
            }
        }
    }

    private val pageSize = 20

    init {
        loadNotifications()
    }

    fun loadNotifications() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null, currentPage = 1) }

            try {
                val response = httpClient.get("/notifications") {
                    parameter("page", 1)
                    parameter("limit", pageSize)
                    if (_uiState.value.filter == NotificationFilter.Unread) {
                        parameter("unread_only", true)
                    }
                }.body<NotificationListResponse>()

                _uiState.update {
                    it.copy(
                        notifications = response.notifications,
                        unreadCount = response.unreadCount,
                        isLoading = false,
                        isRefreshing = false,
                        error = null,
                        currentPage = 1,
                        hasMore = response.notifications.size >= pageSize
                    )
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        isRefreshing = false,
                        error = e.message ?: "Failed to load notifications"
                    )
                }
            }
        }
    }

    fun loadMore() {
        val state = _uiState.value
        if (state.isLoadingMore || !state.hasMore) return

        viewModelScope.launch {
            val nextPage = state.currentPage + 1
            _uiState.update { it.copy(isLoadingMore = true) }

            try {
                val response = httpClient.get("/notifications") {
                    parameter("page", nextPage)
                    parameter("limit", pageSize)
                    if (state.filter == NotificationFilter.Unread) {
                        parameter("unread_only", true)
                    }
                }.body<NotificationListResponse>()

                _uiState.update {
                    it.copy(
                        notifications = it.notifications + response.notifications,
                        unreadCount = response.unreadCount,
                        isLoadingMore = false,
                        currentPage = nextPage,
                        hasMore = response.notifications.size >= pageSize
                    )
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(isLoadingMore = false) }
            }
        }
    }

    fun refresh() {
        _uiState.update { it.copy(isRefreshing = true) }
        loadNotifications()
    }

    fun setFilter(filter: NotificationFilter) {
        if (_uiState.value.filter == filter) return
        _uiState.update { it.copy(filter = filter) }
        loadNotifications()
    }

    fun markRead(notificationId: String) {
        viewModelScope.launch {
            try {
                httpClient.post("/notifications/$notificationId/read")

                _uiState.update { state ->
                    val updated = state.notifications.map { notification ->
                        if (notification.id == notificationId) {
                            notification.copy(read = true)
                        } else {
                            notification
                        }
                    }
                    state.copy(
                        notifications = updated,
                        unreadCount = (state.unreadCount - 1).coerceAtLeast(0)
                    )
                }
            } catch (_: Exception) {
                // Silently fail - will sync on next refresh
            }
        }
    }

    fun markAllRead() {
        viewModelScope.launch {
            val unreadIds = _uiState.value.notifications
                .filter { !it.read }
                .map { it.id }

            // Optimistic update
            _uiState.update { state ->
                state.copy(
                    notifications = state.notifications.map { it.copy(read = true) },
                    unreadCount = 0
                )
            }

            // Send to server
            for (id in unreadIds) {
                try {
                    httpClient.post("/notifications/$id/read")
                } catch (_: Exception) {
                    // Continue with remaining
                }
            }
        }
    }
}
